### ISSUE TITLE
Loading a SavedModel crashes Java

### ISSUE THREAD

**Issue by jeisinge** [tensorflow/java#194]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.3.1
- Python version: 3.8

**Describe the current behavior**
When I load a SavedModel, the JVM crashes.

**Describe the expected behavior**
The JVM should not crash.

**Code to reproduce the issue**
I can load other SavedModels.  So, I expect this is something specific to my SavedModel.  Since the model is proprietary, I cannot share.  And, I don't know how to easily triage the issue.  If you can help, please provide some information!

**Other info / logs**
See [hs_err_pid30392.log](https://github.com/tensorflow/java/files/5862925/hs_err_pid30392.log).

**Comment by Craigacp** [comment]
Can you say anything about the overall shape of the model, what kind of ops it uses, if there are any unbound dimensions?

**Comment by Craigacp** [comment]
And this model loads fine in TF Python 2.3.1 right?

**Comment by jeisinge** [comment]
Thank you for the very quick response!

The model does *not* load fine with TF Python 2.3.1.  We were only able to get it to load when we set `compile=false` on `keras.models.load_model` .  `tf.saved_model.load` also fails.  It *does* load fine on TF-Serving.

As for your questions about the model:
 * It's a Keras model. with a lot of internal keras models.
 * The input is all known shapes except for the batch size.
 * The output is all known shapes except for the batch size.
 * To my knowledge, everything internally has known shapes.
 * For operations, if it is possible to list them from a keras model, let me know.
 * A couple of operations/layers that might be relatively unique/weird shapes are dot, ones_like, logical_*, is_finite, where, batch normalization, squeeze, zeros, boolean_mask, stack
 * The model also uses Keras lambda layers to work around model serialization issues.
 * Also, the models uses keras.backend.int_shape

I hope this helps!

**Comment by Craigacp** [comment]
Ok. If it doesn't load fine with TF Python then you should open an issue on the https://github.com/tensorflow/tensorflow repository as the problem is inside the core TF library and we won't be able to fix it on the Java side because we don't have control that far down inside the C++ library.

**Comment by jeisinge** [comment]
I will do that!

With that being said, why does TF-Serving load it fine?  And, why does Keras load it fine (with `compile=False`)?  Might that be a clue to the root cause?

**Comment by jeisinge** [comment]
I created https://github.com/tensorflow/tensorflow/issues/46644 .

**Comment by Craigacp** [comment]
> I will do that!
> 
> With that being said, why does TF-Serving load it fine? And, why does Keras load it fine (with `compile=False`)? Might that be a clue to the root cause?

It probably is, but the model loading code is so far underneath the Java API that I've not looked at it so I can't say what it's telling us. I'd guess some distinction between tf.function and eager mode, but I don't know.

**Comment by jeisinge** [comment]
Thanks for the explanation. I'll see if the other ticket can get resolved.
